src/compiler/checker.ts

```diff
@@ -6386,6 +6386,8 @@ namespace ts {
             // Only narrow when symbol is variable of type any or an object, union, or type parameter type
             if (node && symbol.flags & SymbolFlags.Variable) {
                 if (isTypeAny(type) || type.flags & (TypeFlags.ObjectType | TypeFlags.Union | TypeFlags.TypeParameter)) {
+                    const declaration = getDeclarationOfKind(symbol, SyntaxKind.VariableDeclaration);

```

Comment:
Current changes vs master baseline (5 iterations of perf test):
```
Project                         Baseline (ms)     Current (ms)      Delta (%) Best (ms) Worst (ms)
-------                         -------------     ------------      --------- --------- ----------
Compiler - tsc (ParseTime)      0.73 (± 46.94 %)  0.60 (± 2.76 %)   -17.71 %       0.59       0.62
Compiler - tsc (BindTime)       0.24 (± 4.36 %)   0.24 (± 2.88 %)   -0.84 %        0.23       0.24
Compiler - tsc (CheckTime)      1.11 (± 2.55 %)   1.07 (± 1.32 %)   -3.42 %        1.06       1.09
Compiler - tsc (EmitTime)       2.02 (± 12.41 %)  1.92 (± 7.18 %)   -4.94 %        1.73       1.99
Compiler - tsc (TotalTime)      4.11 (± 13.65 %)  3.83 (± 3.09 %)   -6.62 %        3.67        3.9
Monaco - tsc (ParseTime)        2.47 (± 70.11 %)  1.70 (± 1.11 %)   -31.45 %       1.68       1.72
Monaco - tsc (BindTime)         0.70 (± 2.67 %)   0.66 (± 1.69 %)   -6.82 %        0.65       0.67
Monaco - tsc (CheckTime)        3.35 (± 12.81 %)  3.22 (± 11.00 %)  -3.88 %        2.99       3.57
Monaco - tsc (EmitTime)         6.61 (± 22.60 %)  6.30 (± 15.68 %)  -4.69 %        5.55       7.22
Monaco - tsc (TotalTime)        13.14 (± 18.82 %) 11.87 (± 11.16 %) -9.62 %       10.93      13.14
TFS - tsc (ParseTime)           1.46 (± 70.93 %)  1.08 (± 1.41 %)   -26.23 %       1.06       1.09
TFS - tsc (BindTime)            0.49 (± 14.66 %)  0.46 (± 1.47 %)   -4.92 %        0.46       0.47
TFS - tsc (CheckTime)           2.28 (± 1.83 %)   2.45 (± 6.33 %)   7.73 %         2.29       2.57
TFS - tsc (EmitTime)            3.65 (± 5.27 %)   3.85 (± 22.65 %)  5.59 %         3.49       5.11
TFS - tsc (TotalTime)           7.88 (± 16.79 %)  7.85 (± 12.08 %)  -0.41 %        7.36        9.2
Encyclopedia - tsc (ParseTime)  0.42 (± 22.49 %)  0.38 (± 1.81 %)   -9.62 %        0.37       0.38
Encyclopedia - tsc (BindTime)   0.14 (± 9.58 %)   0.14 (± 7.71 %)   1.41 %         0.13       0.15
Encyclopedia - tsc (CheckTime)  0.50 (± 1.11 %)   0.51 (± 4.71 %)   3.21 %          0.5       0.54
Encyclopedia - tsc (EmitTime)   0.08 (± 28.47 %)  0.08 (± 28.47 %)  0.00 %         0.07       0.11
Encyclopedia - tsc (TotalTime)  1.13 (± 10.39 %)  1.11 (± 3.40 %)   -2.12 %        1.08       1.15
Compiler - node (ParseTime)     0.78 (± 8.57 %)   0.75 (± 1.81 %)   -4.08 %        0.74       0.76
Compiler - node (BindTime)      0.39 (± 10.07 %)  0.36 (± 2.90 %)   -8.21 %        0.35       0.37
Compiler - node (CheckTime)     1.70 (± 10.04 %)  1.64 (± 3.80 %)   -3.30 %        1.57       1.69
Compiler - node (EmitTime)      1.97 (± 4.16 %)   1.90 (± 11.51 %)  -3.26 %        1.74       2.18
Compiler - node (TotalTime)     4.84 (± 6.30 %)   4.66 (± 5.44 %)   -3.84 %        4.42       4.98
Monaco - node (ParseTime)       1.92 (± 8.11 %)   1.90 (± 7.76 %)   -1.14 %        1.78       2.08
Monaco - node (BindTime)        0.99 (± 7.98 %)   0.95 (± 6.98 %)   -4.23 %        0.89       1.03
Monaco - node (CheckTime)       5.59 (± 4.20 %)   5.60 (± 4.81 %)   0.07 %         5.32       5.88
Monaco - node (EmitTime)        5.96 (± 1.06 %)   5.87 (± 5.12 %)   -1.48 %        5.62       6.21
Monaco - node (TotalTime)       14.47 (± 2.89 %)  14.32 (± 4.56 %)  -1.05 %        13.6      14.81
TFS - node (ParseTime)          1.33 (± 4.01 %)   1.32 (± 3.88 %)   -0.45 %        1.28       1.39
TFS - node (BindTime)           0.83 (± 2.36 %)   0.84 (± 2.25 %)   0.72 %         0.82       0.86
TFS - node (CheckTime)          3.75 (± 2.66 %)   3.77 (± 2.40 %)   0.53 %         3.65       3.84
TFS - node (EmitTime)           4.13 (± 4.28 %)   4.20 (± 2.82 %)   1.55 %         4.13       4.36
TFS - node (TotalTime)          10.04 (± 2.89 %)  10.13 (± 2.27 %)  0.82 %         9.98      10.45
Encyclopedia - node (ParseTime) 0.29 (± 1.93 %)   0.30 (± 1.86 %)   3.47 %         0.29        0.3
Encyclopedia - node (BindTime)  0.22 (± 2.50 %)   0.23 (± 4.75 %)   5.41 %         0.22       0.24
Encyclopedia - node (CheckTime) 1.03 (± 2.41 %)   1.08 (± 3.40 %)   4.66 %         1.03        1.1
Encyclopedia - node (EmitTime)  0.13 (± 13.22 %)  0.13 (± 12.65 %)  1.59 %         0.12       0.15
Encyclopedia - node (TotalTime) 1.67 (± 1.93 %)   1.73 (± 2.17 %)   3.84 %         1.68       1.75
```

It causes minimal/no changes to overall perf in real-world scenarios. (Though I imagine this is the kind of change were you'd only see perf changes in a micro-benchmark anyway)